In [1]:
%pip install -q --disable-pip-version-check pandas
%pip install -q --disable-pip-version-check scikit-learn

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [ ]:
# Check the current working directory
import os
os.getcwd()
from pathlib import Path
from sklearn.datasets import fetch_california_housing
# Standard way to import pandas with alias 'pd'
import pandas as pd
import numpy as np

# 1. Which command is the standard convention for importing the Pandas library?

**Answer:** `import pandas as pd`

In [4]:
# Verify that the housing.csv file exists in the expected path
path = r"C:\Users\Elahe\Desktop\AI-Engineering-Lab\shirafkan-Machine-learning\ML1\data\housing.csv"
print("File exists?", os.path.exists(path))
print("Current working directory:", os.getcwd())


File exists? True
Current working directory: c:\Users\Elahe\Desktop\AI-Engineering-Lab\shirafkan-Machine-learning\ML1\Notebook


In [5]:
# This cell demonstrates how to add the target variable to DataFrame
# (See Cell 6 for actual implementation)
# Example: df2['Target'] = data.target

In [6]:
# Load the California housing dataset from scikit-learn
# Install scikit-learn if not already installed
# Lade den California Housing Datensatz
bunch = fetch_california_housing(as_frame=False)  # type: ignore

# Erstelle DataFrame aus Features
df2 = pd.DataFrame(bunch.data, columns=bunch.feature_names)  # type: ignore[union-attr]

# Extrahiere die Zielvariable
target = bunch.target  # type: ignore[union-attr]

# Jetzt kannst du mit target arbeiten
print("Target shape:", target.shape)
print("First 5 target values:")
print(target[:5])

# Füge target als Spalte zum DataFrame hinzu
df2['Target'] = target

# Überprüfe den DataFrame
print("\nDataFrame mit Target-Spalte:")
print(df2.head())

Target shape: (20640,)
First 5 target values:
[4.526 3.585 3.521 3.413 3.422]

DataFrame mit Target-Spalte:
   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  Target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422  


## Save the CSV file to the project

This code saves the DataFrame as a CSV file in the project's data folder so you have a local copy.

In [10]:
# Save the DataFrame to CSV file in the data directory
# ...existing code...


# df2 nur erstellen, falls es noch nicht existiert oder kein DataFrame ist
if "df2" not in globals() or not isinstance(df2, pd.DataFrame):
    housing_result = fetch_california_housing(as_frame=True)

    # Handle both possible return types: Bunch or (data, target) tuple
    if isinstance(housing_result, tuple):
        X, y = housing_result
        df2 = X.copy()
        df2["Target"] = y
    elif hasattr(housing_result, "frame") and housing_result.frame is not None:
        df2 = housing_result.frame.rename(columns={"MedHouseVal": "Target"})
    elif hasattr(housing_result, "data") and hasattr(housing_result, "target"):
        df2 = pd.DataFrame(housing_result.data, columns=housing_result.feature_names)
        df2["Target"] = housing_result.target
    else:
        raise TypeError("Unsupported return type from fetch_california_housing")

# Type guard for static checkers and runtime safety
assert isinstance(df2, pd.DataFrame), "df2 must be a pandas DataFrame"

out_path = Path("..") / "data" / "housing.csv"
out_path.parent.mkdir(parents=True, exist_ok=True)

df2.to_csv(out_path, index=False)
print("File saved to:", out_path.resolve())
# ...existing code...

File saved to: C:\Users\Elahe\Desktop\AI-Engineering-Lab\shirafkan-Machine-learning\ML1\Data\housing.csv


## 2. If the data is stored in a file named "housing.csv", how do you load it into df2?

**Answer:** `df2 = pd.read_csv("../data/housing.csv")`

Use `pd.read_csv()` to load CSV files into a DataFrame.

In [11]:
# Load data from CSV file 
import pandas as pd
df2 = pd.read_csv("../data/housing.csv")

# Display first 5 rows to verify the data was loaded correctly
df2.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


## 3. How can you check the number of rows and columns in df2?

**Answer:** Use `df2.shape` - returns a tuple (rows, columns)

In [12]:
# Check the shape of the DataFrame (number of rows and columns)
import pandas as pd
df2 = pd.read_csv("../data/housing.csv")

# Get the dimensions of the DataFrame
rows, cols = df2.shape
print("Number of rows:", rows)
print("Number of columns:", cols)

Number of rows: 20640
Number of columns: 9


## 4. You want to quickly inspect the first 5 rows to verify the data was loaded correctly

**Answer:** Use `df2.head()`

This displays the first 5 rows of the DataFrame.

In [13]:
# Display the first 5 rows of the DataFrame
df2.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


## 5. Which method provides a summary of the DataFrame including data types and non-null values?

**Answer:** Use `df2.info()`

This shows column names, data types (Dtype), non-null counts, and memory usage.

In [14]:
# Display detailed information about the DataFrame
import pandas as pd
from sklearn.datasets import fetch_california_housing

# Load data
# Lade den Datensatz als Bunch-Objekt
bunch = fetch_california_housing(as_frame=False)

# Jetzt kannst du auf feature_names zugreifen
feature_names = bunch.feature_names
print("Feature names:", feature_names)

# Create DataFrame from features
df2 = pd.DataFrame(bunch.data, columns=bunch.feature_names)
df2['Target'] = bunch.target

# Save to CSV
df2.to_csv("../data/housing_with_target.csv", index=False)

# Show info: column names, data types, non-null counts
df2.info()


Feature names: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']
<class 'pandas.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Latitude    20640 non-null  float64
 7   Longitude   20640 non-null  float64
 8   Target      20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


## 6. To get statistical summaries like mean, standard deviation, and quartiles for numeric columns, which method is used?

**Answer:** Use `df2.describe()`

Returns count, mean, standard deviation, min, 25%, 50%, 75%, and max values.

In [15]:
# Remove duplicate columns (keep only the first occurrence of each column name)
# This ensures each column name appears only once in the DataFrame
# Note: df2 must be defined in a previous cell before running this cell

df2 = df2.loc[:, ~df2.columns.duplicated()]
print("Duplicate columns removed. Current columns:", list(df2.columns))


Duplicate columns removed. Current columns: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude', 'Target']


In [16]:
# Display statistical summary of numeric columns
# Shows: count, mean, std, min, 25%, 50%, 75%, max
df2.describe()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704,2.068558
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532,1.153956
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000,0.149990
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000,1.196000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000,1.797000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000,2.647250
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000,5.000010


## 7. How do you select the 'Population' column as a Pandas Series?

**Answer:** `df2['Population']` or `df2[['Population']]`

Single square brackets return a Series, double square brackets return a DataFrame.

In [17]:
# Select a single column as a Pandas Series
import pandas as pd
df2 = pd.read_csv("../data/housing.csv")

# Display first few rows

print("First few rows:")
print(df2.head())

print("\nPopulation column:")
# Extract the 'Population' column as a Series
print(df2['Population'])

First few rows:
   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  Target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422  

Population column:
0         322.0
1        2401.0
2         496.0
3         558.0
4         565.0
          ...  
20635     845.0
20636     356.0
20637    1007.0
20638     741.0
20639    1387.0
Name: Population, Length: 20640, dtype: float64


In [18]:
# Extract the target variable as a Series
y = df2['Target']

# Check the shape and display first few values
print("Target variable shape:", y.shape)
y.head()

Target variable shape: (20640,)


0    4.526
1    3.585
2    3.521
3    3.413
4    3.422
Name: Target, dtype: float64

In [19]:
# Display all column names as a Python list
print("Column names in the DataFrame:")
print(list(df2.columns))

Column names in the DataFrame:
['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude', 'Target']


## 8. You want to see the last 10 rows of the DataFrame. Which command is correct?

**Answer:** `df2.tail(10)`

The `tail()` method returns the last N rows (default is 5).

In [ ]:
# Display the last 10 rows of the DataFrame
df2 = pd.read_csv("../data/housing.csv")
# Display the last 10 rows
df2.tail(10)

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
20630,3.5673,11.0,5.932584,1.134831,1257.0,2.824719,39.29,-121.32,1.120
20631,3.5179,15.0,6.145833,1.141204,1200.0,2.777778,39.33,-121.40,1.072
20632,3.1250,15.0,6.023377,1.080519,1047.0,2.719481,39.26,-121.45,1.156
20633,2.5495,27.0,5.445026,1.078534,1082.0,2.832461,39.19,-121.53,0.983
20634,3.7125,28.0,6.779070,1.148256,1041.0,3.026163,39.27,-121.56,1.168
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847
20639,2.3886,16.0,5.254717,1.162264,1387.0,2.616981,39.37,-121.24,0.894


## 9. What is the return type of df2.shape?

**Answer:** **Tuple** (rows, columns)

Example: `(20640, 8)` means 20640 rows and 8 columns

In [ ]:
# Get the shape of the DataFrame (returns a tuple)
df2 = pd.read_csv("../data/housing.csv")

# Returns tuple: (number_of_rows, number_of_columns)
print(df2.shape)

(20640, 9)


## 10. Which syntax uses labels (column/index names) to select data?

**Answer:** `df2.loc[row_label, column_label]`

`loc` = label-based indexing (uses column and row names)

In [20]:
# Use .loc to select data by label (row and column names)
import pandas as pd
df2 = pd.read_csv("../data/housing.csv")

# Select row with index 0, column named 'Target' using label-based indexing
df2.loc[0, 'Target']

np.float64(4.526)

## 11. Which syntax uses integer positions to select the first 5 rows and the first 3 columns?

**Answer:** `df2.iloc[:5, :3]`

`iloc` = integer location-based indexing (uses 0-based positions)

In [21]:
# Use .iloc to select data by integer position (0-based indexing)
import pandas as pd
df2 = pd.read_csv("../data/housing.csv")

# Select rows 0-4 (first 5 rows) and columns 0-2 (first 3 columns)
# Note: The end index is excluded (0:5 means indices 0,1,2,3,4)
df2.iloc[:5, :3]

,MedInc,HouseAge,AveRooms
0,8.3252,41.0,6.984127
1,8.3014,21.0,6.238137
2,7.2574,52.0,8.288136
3,5.6431,52.0,5.817352
4,3.8462,52.0,6.281853


In [22]:
# Alternative syntax: explicitly specify the range
# Same result as above
df2.iloc[0:5, 0:3]

,MedInc,HouseAge,AveRooms
0,8.3252,41.0,6.984127
1,8.3014,21.0,6.238137
2,7.2574,52.0,8.288136
3,5.6431,52.0,5.817352
4,3.8462,52.0,6.281853


## Summary: .loc vs .iloc

| Goal | Syntax | Example |
|------|--------|---------|
| Select by **labels** (column/row names) | `df.loc[...]` | `df.loc[0, 'MedHouseVal']` |
| Select by **integer positions** | `df.iloc[...]` | `df.iloc[0, 7]` |
| Select first 5 rows, first 3 columns | `df.iloc[0:5, 0:3]` | Returns rows 0–4 and columns 0–2 |

---

### Key Rules
- **`.loc` uses labels** (row/column names) - END IS INCLUSIVE
- **`.iloc` uses integer positions** (0-based) - END IS EXCLUSIVE

---

### Example DataFrame
```
df2.columns
# Output: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 
#          'Population', 'AveOccup', 'Latitude', 'Longitude', 'MedHouseVal']
```

### Examples
```python
df2.loc[0, 'MedHouseVal']     # Get value at row 0, column 'MedHouseVal' (by label)
df2.iloc[0, 8]               # Get value at row index 0, column index 8 (by position)
df2.iloc[0:5, 0:3]           # Get rows 0-4, columns 0-2 (by position)
```

---

### Exam Tip ✅
> **Use `.loc` for labels, use `.iloc` for integer positions.**

## 12. You want to filter rows where the house price ('Target') is greater than 5. Which code is correct?

**Answer:** `df2[df2['Target'] > 5]`

This uses **Boolean indexing** - create a condition and use it to filter rows.

In [ ]:
# Filter rows where Target value is greater than 5
df2 = pd.read_csv("../data/housing.csv")
# Boolean indexing: select rows where Target > 5
df2[df2['Target'] > 5]

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
89,1.2434,52.0,2.929412,0.917647,396.0,4.658824,37.80,-122.27,5.00001
459,1.1696,52.0,2.436000,0.944000,1349.0,5.396000,37.87,-122.25,5.00001
493,7.8521,52.0,7.794393,1.051402,517.0,2.415888,37.86,-122.24,5.00001
494,9.3959,52.0,7.512097,0.955645,1366.0,2.754032,37.85,-122.24,5.00001
509,7.8772,52.0,8.282548,1.049861,947.0,2.623269,37.83,-122.23,5.00001
...,...,...,...,...,...,...,...,...,...
20422,5.1457,35.0,6.958333,1.217593,576.0,2.666667,34.14,-118.90,5.00001
20426,10.0472,11.0,9.890756,1.159664,415.0,3.487395,34.18,-118.69,5.00001
20427,8.6499,4.0,7.236059,1.032528,5495.0,2.553439,34.19,-118.80,5.00001
20436,12.5420,10.0,9.873315,1.102426,1179.0,3.177898,34.21,-118.69,5.00001


## 13. When combining two conditions (e.g., Target > 200,000 AND HouseAge < 10), which operator must be used inside the DataFrame selection?

**Options:**
- A) `and` ❌ (doesn't work with DataFrames)
- B) `&&` ❌ (Python syntax error)
- C) `&` ✅ (CORRECT - boolean AND operator)
- D) `+` ❌ (addition operator)

**Answer: C) Use the `&` operator**

Example: `df2[(df2['Target'] > 200000) & (df2['HouseAge'] < 10)]`

**Remember:** Always use parentheses around each condition!

## 14. Which of the following is the correct syntax for an "OR" condition in Pandas filtering?

**Answer: B) `df2[(df2['A'] > 1) | (df2['B'] < 2)]`**

Use the `|` (pipe) operator for OR conditions in pandas filtering.

**Important:**
- `&` for AND conditions
- `|` for OR conditions
- Always use **parentheses** around each condition

In [24]:
# Filter rows using OR condition: Target > 1 OR Target < 2
# Use the | (pipe) operator for OR in pandas
df2[(df2['Target'] > 1) | (df2['Target'] < 2)]

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847


## 15. How do you count the number of missing values (NaN) in each column?

**Answer: B) `df2.isna().sum()`**

This returns the count of missing values for each column.

**Alternative:** `df2.isnull().sum()` (does the same thing)

In [25]:
# Count the total number of missing (NaN) values in each column
# Returns a Series with counts per column
df2.isna().sum()

MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
Target        0
dtype: int64

In [ ]:
# Example: Create a DataFrame with some missing values
df2 = pd.DataFrame({
    'A': [1, np.nan, 3],
    'B': [np.nan, np.nan, 4]
})

# Count missing values in each column
df2.isna().sum()

A    1
B    2
dtype: int64

## 16. Which command will remove all rows from df2 that contain at least one missing value?

**Answer: C) `df2.dropna()`**

Removes any row that has at least one NaN (missing) value.

**Note:** Default behavior is to drop rows axis=0), not columns.

In [27]:
# Remove all rows containing at least one missing value (NaN)
# Returns a new DataFrame without rows that have any NaN
df2_clean = df2.dropna()

# Show comparison
print(f"Rows before: {len(df2)}")
print(f"Rows after: {len(df2_clean)}")

Rows before: 3
Rows after: 1


In [ ]:
# Example: Create DataFrame with missing values and remove them
df2 = pd.DataFrame({
    'A': [1, np.nan, 3],
    'B': [4, 5, np.nan]
})

print("Original DataFrame:")
print(df2)
print("\nAfter dropna():")
print(df2.dropna())

Original DataFrame:
     A    B
0  1.0  4.0
1  NaN  5.0
2  3.0  NaN

After dropna():
     A    B
0  1.0  4.0


## Understanding `df2.dropna()`

`df2.dropna()` means:

> **Remove any row that contains at least one missing value (NaN).**

### Simple Example

Suppose we have this DataFrame:

| A   | B   |
| --- | --- |
| 1   | 4   |
| NaN | 5   |
| 3   | NaN |

If we run: `df2.dropna()`

### What happens?

* Row 2 ❌ (because A is missing)
* Row 3 ❌ (because B is missing)
* Only Row 1 remains ✅

Result:

| A | B |
| - | - |
| 1 | 4 |

---

### Important Note

* `dropna()` works on **rows by default**
* If **even one NaN** exists in a row, that row is removed

---

### Alternative: Count missing values instead of removing

```python
df2.isna().sum()
```

This counts how many missing values are in each column **without removing rows**.

---

### Quick Summary

> **`df2.dropna()` = Remove all rows that contain at least one missing value**


## 17. Instead of removing data, you want to replace all missing values in the 'Target' column with the mean price

**Answer:** Use `df2['Target'] = df2['Target'].fillna(df2['Target'].mean())`

The `fillna()` method replaces NaN values with the specified value (in this case, the mean).

In [ ]:
# Load the data
df2 = pd.read_csv("../data/housing.csv")

## Method: Fill missing values in Target column with the mean

To replace missing values in the Target column with its average value:

In [31]:
# Fill missing values in Target column with the mean of the column
# Step 1: Calculate the mean value
mean_target = df2["Target"].mean()

# Step 2: Replace all NaN with that mean
df2["Target"] = df2["Target"].fillna(mean_target)

# Verify no missing values remain
print(f"Missing values after filling: {df2['Target'].isna().sum()}")

Missing values after filling: 0


In [ ]:
# Demonstrate the fillna() method with a simple example
# Create a Series with some NaN values
s = pd.Series([1, np.nan, 3, np.nan, 5])
print("Original Series:")
print(s)

# Fill NaN with the mean
print("\nAfter fillna(mean):")
print(s.fillna(s.mean()))

Original Series:
0    1.0
1    NaN
2    3.0
3    NaN
4    5.0
dtype: float64

After fillna(mean):
0    1.0
1    3.0
2    3.0
3    3.0
4    5.0
dtype: float64


In [34]:
# If MedHouseVal column exists, create Target column from it
if "MedHouseVal" in df2.columns and "Target" not in df2.columns:
    df2["Target"] = df2["MedHouseVal"]

# Fill any missing values in Target with the column mean
if "Target" in df2.columns:
    df2["Target"] = df2["Target"].fillna(df2["Target"].mean())
    print("Target column has been processed successfully")

Target column has been processed successfully


In [36]:
# Check if Target column exists
print("Target in columns:", "Target" in df2.columns)

Target in columns: True


## Important Note: `inplace=True` vs Direct Assignment

**Old approach (deprecated):**
```python
df2["Target"].fillna(df2["Target"].mean(), inplace=True)
```

**Modern approach (recommended):**
```python
df2["Target"] = df2["Target"].fillna(df2["Target"].mean())
```

✅ **Direct assignment is safer and more future-proof** than using `inplace=True`

## 18. You suspect there are duplicate rows in the dataset. How do you remove them?

**Answer:** `df2 = df2.drop_duplicates()`

Removes all duplicate rows, keeping only the first occurrence of each unique row.

In [ ]:
# Remove duplicate rows from the DataFrame
# Only keeps the first occurrence of each duplicate row
df2 = pd.read_csv("../data/housing.csv")

# Drop duplicate rows
df2 = df2.drop_duplicates()
df2.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


In [38]:
# Count rows before and after removing duplicates
before = len(df2)
df2 = df2.drop_duplicates()
after = len(df2)

# Show how many duplicate rows were removed
print(f"Duplicate rows removed: {before - after}")
print(f"Total rows remaining: {after}")

Duplicate rows removed: 0
Total rows remaining: 20640


## 19. You want to rename the column 'Target' to 'MedianHouseValue'. Which syntax is correct?

**Answer:** `df2 = df2.rename(columns={'Target': 'MedianHouseValue'})`

The `rename()` method with the `columns` parameter changes column names.

In [39]:
# Rename a column in the DataFrame
# Old column name 'Target' becomes 'MedianHouseValue'
df2 = df2.rename(columns={'Target': 'MedianHouseValue'})

print("Column renamed successfully")
print("New column names:", list(df2.columns))

Column renamed successfully
New column names: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude', 'MedianHouseValue']


## 20. If there is a categorical column named 'Ocean_Proximity', how can you see the count of each unique category?

**Answer: C) `df['Ocean_Proximity'].value_counts()`**

`value_counts()` shows the count of each unique value in a column, sorted in descending order.

In [40]:
# Display the count of unique values in a column
# Sorted from most frequent to least frequent
df2['Latitude'].value_counts()

Latitude
34.06    244
34.05    236
34.08    234
34.07    231
34.04    221
        ... 
35.97      1
35.86      1
35.85      1
35.89      1
35.88      1
Name: count, Length: 862, dtype: int64

## 21. How do you sort the DataFrame based on the 'Target' column in descending order (highest price first)?

**Answer: A) `df.sort_values(by='Target', ascending=False)`**

- `by='Target'` - specify the column to sort by
- `ascending=False` - sort in descending order (highest first)
- `ascending=True` - sort in ascending order (lowest first, default)

In [ ]:
# Sort DataFrame by Target column in descending order (highest values first)
df2 = pd.read_csv("../data/housing.csv")

# If Target doesn't exist, create it from MedHouseVal
if "Target" not in df2.columns and "MedHouseVal" in df2.columns:
    df2["Target"] = df2["MedHouseVal"]

# Sort by Target descending
df2.sort_values(by="Target", ascending=False)

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,Target
5253,13.2935,27.0,7.607143,1.012673,2336.0,2.691244,34.11,-118.49,5.00001
5254,10.7937,29.0,7.471787,1.217868,1500.0,2.351097,34.07,-118.48,5.00001
5255,8.5153,40.0,6.407266,0.925430,1564.0,2.990440,34.07,-118.48,5.00001
5256,12.8665,37.0,7.457565,1.012915,1318.0,2.431734,34.07,-118.48,5.00001
5257,15.0001,42.0,9.229032,1.161290,829.0,2.674194,34.06,-118.49,5.00001
...,...,...,...,...,...,...,...,...,...
5887,2.3667,39.0,3.572464,1.217391,259.0,1.876812,34.15,-118.33,0.17500
9188,4.1932,52.0,3.568889,1.186667,628.0,2.791111,34.24,-117.86,0.14999
19802,0.5360,36.0,12.250000,3.500000,18.0,2.250000,40.31,-123.17,0.14999
2799,2.1000,19.0,3.774390,1.457317,490.0,2.987805,36.40,-117.02,0.14999


In [ ]:
# Sort DataFrame by Target column in descending order
df2.sort_values(by='Target', ascending=False)

## 22. What happens if you run df['Target']?

**Answer: B) It returns a Pandas Series.**

A Series is a 1-D labeled array that is one column from the DataFrame.

- Single square brackets `df['Target']` → Returns a **Series**
- Double square brackets `df[['Target']]` → Returns a **DataFrame**

In [ ]:
# Selecting a single column returns a Pandas Series
# A Series is a 1-dimensional labeled array
df2['Target']

## 23. You want to select rows with index labels 10 to 20 (inclusive) and the column named 'Target'

**Answer: B) `df.loc[10:20, 'Target']`**

- `.loc` uses **label-based** indexing
- When using `.loc`, the end label is **inclusive** (unlike `.iloc`)
- So `10:20` includes both rows 10 and 20

In [ ]:
# Select rows 10-20 (inclusive) and only the Target column using .loc
# .loc is label-based and the end is INCLUSIVE
df2.loc[10:20, 'Target']

## 24. Which attribute would you use to find the data type of every column?

**Answer: A) `df.dtypes`**

Returns a Series showing the data type (dtype) of each column.

Example output:
```
MedInc       float64
HouseAge     float64
AveRooms     float64
MedHouseVal  float64
```

In [ ]:
# Display the data type of each column
df2.dtypes

## 25. What is the result of df.isna() without .sum()?

**Answer: B) A DataFrame of the same size containing Boolean values (True/False)**

- `df.isna()` → Returns DataFrame with **True** for NaN cells and **False** for non-NaN cells
- `df.isna().sum()` → Returns counts of missing values per column

Example:
```python
df2.isna()  # Returns DataFrame with True/False
df2.isna().sum()  # Returns counts of missing values
```

## 26. You want to sort df by 'HouseAge' (ascending) and then by 'Target' (descending)

**Answer: A) `df.sort_values(by=['HouseAge', 'Target'], ascending=[True, False])`**

Pass lists to `by` and `ascending` parameters to sort by multiple columns with different orders.

In [ ]:
# Sort by multiple columns with different orders
# First by HouseAge (ascending), then by Target (descending)
df2.sort_values(by=['HouseAge', 'Target'], ascending=[True, False])

## 27. Which code creates a new DataFrame df_new containing only 'Target' and 'Population' columns?

**Answer: B) `df_new = df[['Target', 'Population']]`**

Use double square brackets with a list of column names to select multiple columns.

In [ ]:
# Select specific columns to create a new DataFrame
# Use double square brackets with a list of column names
df_new = df2[['Target', 'Population']]
df_new.head()

## 28. What is the default behavior of df.dropna()?

**Answer: C) It drops rows where at least one value is missing.**

Default parameters:
- `axis=0` - drop rows (not columns)
- `how='any'` - drop if ANY value is missing (or use 'all' to drop only if ALL are missing)

## 29. Why is using df.loc or df.iloc preferred over chained indexing (e.g., df['col'][0])?

**Answer: D) All of the above**

Reasons:
- **Clarity** - More readable code
- **Performance** - Slightly faster
- **Avoids SettingWithCopyWarning** - Prevents confusing warnings when modifying data
- **Best Practice** - Recommended by pandas developers

**Recommended approaches:**
1. `.loc` for label-based indexing (by column/row names)
2. `.iloc` for position-based indexing (by integer positions)
3. Avoid chained indexing like `df['col'][0]`

In [ ]:
# Chained indexing (NOT recommended - can cause SettingWithCopyWarning)
value1 = df2['MedInc'][0]
print("Value using chained indexing:", value1)


In [ ]:
# Better approach 1: Use .iloc on the Series
value2 = df2['MedInc'].iloc[0]
print("Value using .iloc on Series:", value2)


In [ ]:
# Best approach 2: Use .loc (label-based indexing - RECOMMENDED)
value3 = df2.loc[0, 'MedInc']
print("Value using .loc (RECOMMENDED):", value3)


In [ ]:
# Best approach 3: Use .iloc (position-based indexing)
value4 = df2.iloc[0, 0]
print("Value using .iloc:", value4)

# Verify all methods return the same value
print("\nAll methods return the same values as shown above")

## 30. If you run df.info(), which of the following is NOT displayed?

**Answer: B) The statistical mean of every column.**

`df.info()` shows:
- ✅ Column names
- ✅ Data types (Dtype)
- ✅ Non-null counts
- ✅ Memory usage
- ❌ **NOT** statistical means

For statistical summaries (mean, std, etc.), use `df.describe()` instead.

In [ ]:
# df.describe() shows statistical summaries (mean, std, etc)
# But NOT data types or non-null counts
df2.describe()

In [ ]:
# Comparison:
# - df.info() shows: dtypes, non-null counts, memory usage
# - df.describe() shows: statistical summaries (mean, std, quartiles, etc.)
# Both are useful for different purposes